src/main/java/org/primefaces/component/spinner/SpinnerRenderer.java

```diff
@@ -124,12 +123,52 @@ protected void encodeMarkup(FacesContext context, Spinner spinner) throws IOExce
 
         encodeInput(context, spinner);
 
-        encodeButton(context, upButtonClass, Spinner.UP_ICON_CLASS);
-        encodeButton(context, downButtonClass, Spinner.DOWN_ICON_CLASS);
+        boolean valid = spinner.isValid();
+        String upButtonClass = getButtonClass(context, Spinner.UP_BUTTON_CLASS, spinner.getUpButtonStyleClass(), valid);
+        String downButtonClass = getButtonClass(context, Spinner.DOWN_BUTTON_CLASS, spinner.getDownButtonStyleClass(), valid);
+
+        boolean stacked = SpinnerBase.BUTTONS_STACKED.equals(spinner.getButtons());
+        String upIconClass = getIconClass(context,
+                spinner.getUpIcon(),
+                stacked ? Spinner.STACKED_UP_ICON_CLASS : Spinner.HORIZONTAL_UP_ICON_CLASS);
+        String downIconClass = getIconClass(context,
+                spinner.getDownIcon(),
+                stacked ? Spinner.STACKED_DOWN_ICON_CLASS : Spinner.HORIZONTAL_DOWN_ICON_CLASS);
+
+        encodeButton(context, upButtonClass, upIconClass);
+        encodeButton(context, downButtonClass, downIconClass);
 
         writer.endElement("span");
     }
 
+    protected String getButtonsClassSuffix(Spinner spinner) {
+        switch (spinner.getButtons()) {
+            case SpinnerBase.BUTTONS_HORIZONTAL:
+                return SpinnerBase.BUTTONS_HORIZONTAL;
+            case SpinnerBase.BUTTONS_HORIZONTAL_AFTER:
+                return SpinnerBase.BUTTONS_HORIZONTAL_AFTER;
+            case SpinnerBase.BUTTONS_VERTICAL:
+                return SpinnerBase.BUTTONS_VERTICAL;
+            default:
+                return SpinnerBase.BUTTONS_STACKED;
+        }
+    }
+
+    protected String getButtonClass(FacesContext context, String fixedStyleClass, String styleClass, boolean valid) {
+        return getStyleClassBuilder(context)
+                .add(fixedStyleClass)
+                .add(styleClass)
+                .add(!valid, "ui-state-error")
+                .build();
+    }
+
+    protected String getIconClass(FacesContext context, String custom, String fallback) {
+        return getStyleClassBuilder(context)
+                .add(Spinner.ICON_BASE_CLASS)
+                .add(Optional.ofNullable(custom).orElse(fallback))
```

Comment:
why not just a simple if? Hope we dont use it in other places

Reply:
@tandraschko Removed the `Optional`. It's not used elsewhere.